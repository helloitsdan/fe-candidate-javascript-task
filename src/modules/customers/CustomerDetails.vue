<template>
  <div>
    <div v-if="!customer" class="container padded-block">
      <div v-if="loading" class="loading">Loading...</div>

      <div v-if="error" class="error">
        Sorry! We couldn't load this customer's details right now.
      </div>
    </div>

    <div v-if="customer">
      <header class="container padded-block">
        <nav>
          <p>
            <router-link to="/">Back to customer list</router-link>
          </p>
        </nav>

        <div class="columns">
          <div class="column is-two-thirds">
            <h1 class="is-marginless">{{ customer.name | formatName }}</h1>
            <p>{{ customer.email }}</p>

            <h3>Contact information</h3>
            <p>{{ customer.phone }}</p>
            <p>{{ customer.address }}</p>
          </div>

          <div class="column is-one-third">
            <p>
              <strong>Registered on</strong>
              {{ customer.registered | moment("dddd, MMMM Do YYYY") }}
            </p>

            <p v-if="!customer.isActive">
              This customer is
              <strong>inactive.</strong>
            </p>
          </div>
        </div>
      </header>

      <router-view></router-view>
    </div>
  </div>
</template>

<script lang="ts">
import Vue from "vue";

import getCustomers from "@/services/customers";

export default Vue.extend({
  name: "CustomerDetails",
  props: ["customerId"],
  data() {
    return {
      loading: false,
      error: null,
      customer: null
    };
  },
  created() {
    this.fetchCustomerDetails();
  },
  methods: {
    async fetchCustomerDetails() {
      const { customerId } = this.$props;

      if (this.loading) {
        return;
      }

      this.loading = true;

      try {
        this.customer = await getCustomers(customerId);
      } catch (e) {
        this.error = e;
      }

      this.loading = false;
    }
  }
});
</script>
