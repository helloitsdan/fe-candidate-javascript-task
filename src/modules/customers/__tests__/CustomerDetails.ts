import Vue from "vue";
import { shallowMount } from "@vue/test-utils";

import CustomerDetails from "../CustomerDetails.vue";
import customerRequest from "@/services/customers";

jest.mock("@/services/customers");

describe("CustomerDetails", () => {
  const customerRequestMock = customerRequest as jest.Mock;

  const fakeCustomer = {
    uuid: "1234",
    isActive: true,
    name: {
      first: "Test",
      last: "Customer"
    },
    email: "test.user@pretend.biz",
    registered: "2016-07-13T02:57:55.235Z"
  };

  it("renders", async () => {
    customerRequestMock.mockReturnValue(Promise.resolve(fakeCustomer));

    const wrapper = shallowMount(CustomerDetails, {
      stubs: ["router-link", "router-view"]
    });

    await Vue.nextTick();

    expect(wrapper.element).toMatchSnapshot();
  });

  it("renders a loading message when the request is still pending", () => {
    const wrapper = shallowMount(CustomerDetails, {
      stubs: ["router-link", "router-view"]
    });
    expect(wrapper.find(".loading").exists()).toBe(true);
  });

  it("renders an error message when the request fails", async () => {
    customerRequestMock.mockReturnValue(Promise.reject(":("));
    const wrapper = shallowMount(CustomerDetails, {
      stubs: ["router-link", "router-view"]
    });

    await Vue.nextTick();

    expect(wrapper.find(".error").exists()).toBe(true);
  });
});
