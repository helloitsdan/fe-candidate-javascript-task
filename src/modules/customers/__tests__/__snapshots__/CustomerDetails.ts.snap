// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomerDetails renders 1`] = `
<div>
  <!---->
   
  <div>
    <header
      class="container padded-block"
    >
      <nav>
        <p>
          <router-link-stub
            to="/"
          >
            Back to customer list
          </router-link-stub>
        </p>
      </nav>
       
      <div
        class="columns"
      >
        <div
          class="column is-two-thirds"
        >
          <h1
            class="is-marginless"
          >
            Test Customer
          </h1>
           
          <p>
            test.user@pretend.biz
          </p>
           
          <h3>
            Contact information
          </h3>
           
          <p>
            
          </p>
           
          <p>
            
          </p>
        </div>
         
        <div
          class="column is-one-third"
        >
          <p>
            <strong>
              Registered on
            </strong>
            
            Wednesday, July 13th 2016
          
          </p>
           
          <!---->
        </div>
      </div>
    </header>
     
    <router-view-stub />
  </div>
</div>
`;
